### PR TITLE
rosdep: Added m2r package

### PR DESCRIPTION
**Pull request by peci1**
Please add the following dependency to the rosdep database.

## Package name:

m2r

## Package Upstream Source:

https://github.com/miyakogi/m2r

## Purpose of using this:

This package allows including Markdown files in RST docs, as e.g. mentioned here: https://stackoverflow.com/a/46012023/1076564 .

## Links to Distribution Packages

- Debian: https://packages.debian.org/bullseye/m2r (buster via pip)
- Ubuntu: https://packages.ubuntu.com/jammy/m2r (bionic via pip)
- Fedora: https://packages.fedoraproject.org/pkgs/python-m2r/python3-m2r/
- Arch: not available ([pip](https://pypi.org/project/m2r/))
- Gentoo: not available ([pip](https://pypi.org/project/m2r/))
- macOS: not available ([pip](https://pypi.org/project/m2r/))
- Alpine: not available ([pip](https://pypi.org/project/m2r/))
- NixOS/nixpkgs: https://search.nixos.org/packages?channel=22.11&show=m2r&from=0&size=50&sort=relevance&type=packages&query=m2r
- openSUSE: https://software.opensuse.org/package/python3-m2r

penSUSE: [python-cbor2](https://software.opensuse.org/package/python-cbor2)

## Checks

- [ ] All packages have a declared license in the package.xml
- [x] This repository has a LICENSE file
- [ ] This package is expected to build on the submitted rosdistro

### CODE REVIEW

**Comment by peci1** [comment]
Solution for the pip version constraint provided in https://github.com/ros-infrastructure/rosdep/pull/909 .

**Review by clalancette — COMMENTED** [review]
Two things here:

1.  This seems to be a python package, so I would have expected this to be in rosdep/python.yaml.
2. The upstream repository is now archived and says "no more updates".  Should we really be adding this in at this time?

**Comment by peci1** [comment]
1. This can be used both as a Python package and as a CLI tool. That's why I've added it in base, but I can move it over to Python if you like.

2. The Ubuntu and Debian packages will of course receive maintenance during the lifetime of the respective distros. The pip package will not. Pip is used for a few distros and most importantly for Python 2.7 distros of Ubuntu and Debian. These will soon lose support, so it might be possible getting in a package that got archived in Nov 2022?
 
Alternatively, there is https://github.com/CrossNox/m2r2 which is a fork updated for newer versions of Sphinx. This could be used for the Python 3.7+ distros, but not for Python 2.7 (bionic) or Python 3.6 (focal). m2r2 says it is API-compatible with m2r, although it might be compatible with slightly different versions of tooling (regarding Sphinx, I'm not sure if it's compatible with Sphinx 1, or only with Sphinx 2).

So, what about the following?

```YAML
m2r:
  alpine:
    pip:
      packages: [m2r2]
  arch:
    pip:
      packages: [m2r2]
  debian:
    '*': [m2r]
    buster:
      pip:
        packages: ['m2r<0.3']
  fedora: [python3-m2r]
  gentoo:
    pip:
      packages: [m2r2]
  nixos: [m2r]
  opensuse: [python3-m2r]
  osx:
    pip:
      packages: [m2r2]
  ubuntu:
    '*': [m2r]
    bionic:
      pip:
        packages: ['m2r<0.3']
```

This would use distro packages for m2r where available, m2r2 on the "rolling" distros which tend to be very up-to-date, and pip m2r < 0.3 for bionic and focal which use Python 2.7 (still pending https://github.com/ros-infrastructure/rosdep/pull/909).

File: rosdep/base.yaml
Comment (nuclearsandwich) [inline rosdep/base.yaml:6624]:
```suggestion
        packages: ['m2r']
```

This form will not error during `rosdep install` but rosdep does not support detecting this as installed. If the "current" version of m2r does not work on Bionic and there is no packaged version, I think the best thing is leaving it undefined.

**Comment by nuclearsandwich** [comment]
@peci1 this turned into something of a big one, thanks for your patience while we review it.

> 1. This can be used both as a Python package and as a CLI tool. That's why I've added it in base, but I can move it over to Python if you like.

This is, I think a bit of a gut-call. I tend to follow Debian's lead, if the Debian package is `python3-m2r` then it's best in python.yaml. But there are other equally valid heuristics like whether the package contains a python module, or whether the key has definitions using pip.
I actually think that any rosdep key with definitions from pip belongs in `python.yaml` by our current practices.



> I think the syntax I used should work, but I'm apparently the first one to do it so I'm rather checking if you think the same.

I flagged this in my review comments before catching up on the conversation that you tested and found the same results (and submitted a PR which potentially fixes it!) yourself. In order to review the rest of this change without a significant review and release on the rosdep side I still advocate removing this form, and thus, leaving this key undefined on bionic.

> Alternatively, there is https://github.com/CrossNox/m2r2 which is a fork updated for newer versions of Sphinx. This could be used for the Python 3.7+ distros, but not for Python 2.7 (bionic) or Python 3.6 (focal). m2r2 says it is API-compatible with m2r, although it might be compatible with slightly different versions of tooling (regarding Sphinx, I'm not sure if it's compatible with Sphinx 1, or only with Sphinx 2).

Unless m2r2 is a drop in replacement which provides a compatible `m2r` module (and thus would likely conflict with the original `m2r` package I do not think they should be mixed in the same key definition and instead `m2r2` should be a separate rosdep key. A package which is specifically set up to use either as appropriate could then declare the correct dependency using package.xml conditions.

**Review by nuclearsandwich — CHANGES_REQUESTED** [review]
One of the key definitions is not compatible with current versions of rosdep and should not be merged at this time. The rest looks good as is.

**Comment by peci1** [comment]
Thanks for the extensive feedback, @nuclearsandwich ;)

> I actually think that any rosdep key with definitions from pip belongs in `python.yaml` by our current practices.

No problem, I'll move it there.

> In order to review the rest of this change without a significant review and release on the rosdep side I still advocate removing this form, and thus, leaving this key undefined on bionic.

One more idea: I could create a rdmanifest file in some of our repos and specify `source` installer instead of `pip`. In the rdmanifest, it could be quite easy to do the pip magic correctly. Would this be acceptable?

> Unless m2r2 is a drop in replacement which provides a compatible `m2r` module

Ah, right, I've missed that there would be a name difference... It is API compatible, but up to this "detail" :) I'll thus make one m2r-only key and one m2r2-only.

**Comment by nuclearsandwich** [comment]
> One more idea: I could create a rdmanifest file in some of our repos and specify `source` installer instead of `pip`. In the rdmanifest, it could be quite easy to do the pip magic correctly. Would this be acceptable?

I am boring and prefer to keep things conventional. But I appreciate the amount of effort being put in to making this dependency available wherever possible even if "unconventional". My advice is: if you're not going to be the one using this dependency on the affected platforms, then leaving it blank for now is the way to go.

If you're working to support these platforms because you need that support there, it's another matter entirely. I would lean toward a solution that above all else: does not require a user running `rosdep install` to do anything else out of the ordinary in order to ensure that the setup will work and secondarily: does not require someone adding a dependency on the rosdep key to do anything unusual in _their_ repository setup to make it work. Which is why adding an rdmanifest to a specific repository doesn't seem like a general solution to me.

In either, case you always have the option of using custom rosdep sources in addition to the upstream sources in your project! Since these are pip/source dependencies any way there is no way to use them in the official ROS package builds.



> Ah, right, I've missed that there would be a name difference... It is API compatible, but up to this "detail" :) I'll thus make one m2r-only key and one m2r2-only.

Yeah, my guiding principle here is that if the source code / package itself has to be aware of the difference between two different dependency solutions then those should be different rosdep keys.

**Comment by peci1** [comment]
> My advice is: if you're not going to be the one using this dependency on the affected platforms, then leaving it blank for now is the way to go.

I actually am the one who wants to use it =) And since Melodic EOL is coming, I'd like to leave the project in a good shape before the EOL.

> does not require a user running `rosdep install` to do anything else out of the ordinary in order to ensure that the setup will work

I have the impression that source installs are handled "out of the box" by rosdep (only the buildfarm ignores them). Or am I wrong?

> does not require someone adding a dependency on the rosdep key to do anything unusual in _their_ repository setup to make it work

What do you mean by unusual? The proposed rdmanifest would just run something like `python2 -m pip install 'm2r<0.3'`. Pip is already installed because of rosdep, so there should really be no annoyances. 

> In either, case you always have the option of using custom rosdep sources in addition to the upstream sources in your project!

Sure.

> Yeah, my guiding principle here is that if the source code / package itself has to be aware of the difference between two different dependency solutions then those should be different rosdep keys.

I agree.

---

You as a maintainer have the last word and I think we've already discussed everything related. So please go on and give a definitive decision whether the bionic/buster keys should be added or not.

**Comment by nuclearsandwich** [comment]
> I have the impression that source installs are handled "out of the box" by rosdep (only the buildfarm ignores them). Or am I wrong?

I have no problem admitting that my own experience is so concentrated on the build farm delivery that I have no idea. All rosdistro PRs require two reviewers and I think it's time for me to call in backup early. @cottsay do you have more experience with the `source` definitions for rosdep? I am at the point where I would have to experiment.

> You as a maintainer have the last word and I think we've already discussed everything related. So please go on and give a definitive decision whether the bionic/buster keys should be added or not.

I appreciate the acknowledgement. I'm definitely not trying to wear you down but also recognize that this has been a long one. My motivation for asking whether you were going to be using the atypical definitions was to figure out whether continuing to iterate was worthwhile. Since you are and you're trying to set Melodic packages up to go out with up-to-date support, I do not want to shut that down prematurely. If the discussion is blocking forward progress elsewhere, we could split the buster and bionic handling into a follow-up PR.

**Comment by cottsay** [comment]
> cottsay, do you have more experience with the source definitions for rosdep?

I have none, sorry :(

**Comment by peci1** [comment]
I think the proposed solution is the best now taking into account Melodic EOL. I'll update the PR.

**Comment by peci1** [comment]
I've updated the PR as suggested.

**Review by nuclearsandwich — APPROVED** [review]

**Review by nuclearsandwich — APPROVED** [review]
Somehow managed to dismiss my own review. It does require a second to merge.

**Review by clalancette — APPROVED** [review]

**Pull request merged**
